### ISSUE TITLE
Raising "list index out of range" error in torch_geometric/data/collate.py

### ISSUE THREAD

**Issue by McDifference** [pyg-team/pytorch_geometric#3802]
## 🐛 Bug

When I used torch_geometric.data.Data as data structure to build an InMemoryDataset, the following error raised in the collate function.
```
File ".../torch_geometric/data/collate.py", line 184, in _collate
    and isinstance(elem[0], (Tensor, SparseTensor))):
IndexError: list index out of range
```

This error also ocurred in 
```
File ".../torch_geometric/data/separate.py", line 92, in _separate
    and isinstance(value[0][0], (Tensor, SparseTensor))):
IndexError: list index out of range
```

### To Reproduce

```
import torch
from torch_geometric.data import InMemoryDataset, download_url
from torch_geometric.data import Data

class MyOwnDataset(InMemoryDataset):
    def __init__(self, root, transform=None, pre_transform=None):
        super().__init__(root, transform, pre_transform)
        self.data, self.slices = torch.load(self.processed_paths[0])

    @property
    def raw_file_names(self):
        return ['some_file_1', 'some_file_2', ...]

    @property
    def processed_file_names(self):
        return ['data.pt']

    def download(self):
        # Download to `self.raw_dir`.
        download_url(url, self.raw_dir)

    def process(self):
        # Read data into huge `Data` list.
        data_list = []
        graph_input = Data(
            x=torch.from_numpy(x).float(),
            y=torch.from_numpy(y).float(),
            cluster=torch.from_numpy(cluster).short(),
            edge_index=torch.from_numpy(edge_index).long(),
            identifier=torch.from_numpy(identifier).float(),
            traj_len=torch.tensor([traj_lens[ind]]).int(),
            valid_len=torch.tensor([valid_lens[ind]]).int(),
            time_step_len=torch.tensor([num_valid_len_max]).int(),
            candidate_len_max=torch.tensor([num_candidate_max]).int(),
            candidate_mask=[],
            candidate=torch.from_numpy(raw_data['tar_candts'].values[0]).float(),
            candidate_gt=torch.from_numpy(raw_data['gt_candts'].values[0]).bool(),
            offset_gt=torch.from_numpy(raw_data['gt_tar_offset'].values[0]).float(),
            target_gt=torch.from_numpy(raw_data['gt_preds'].values[0][0][-1, :]).float(),
        )
        data_list.append(graph_input)
        data, slices = self.collate(data_list)
        torch.save((data, slices), self.processed_paths[0])
```

### Expected behavior

No error raised.

### Environment

* PyG version: torch_geometric.2.0.4
* PyTorch version: torch.1.10.1
* OS: Ubuntu16.04
* Python version: 3.9
* CUDA/cuDNN version: cuda11.3
* How you installed PyTorch and PyG: pip
* Any other relevant information: torch-scatter.2.0.9

**Comment by rusty1s** [comment]
I'm not entirely sure what's going on. I'm sorry. Do you have an executable and minimal example to reproduce this?

**Comment by zmyzxb** [comment]
It's seems that your code is based on a lower version of torch-geometric. I just solved the same problem by uninstalling pyg(from conda) and reinstall torch-geometric=1.7.2 with pip.

**Comment by rusty1s** [comment]
@zmyzxb thanks for chiming in. Does that mean that there is some bug in PyG 2.0.*?

**Comment by alvarogunawan** [comment]
If it helps, I'm encountering the same issue (at least the collate one) when attempting to use a `DataLoader` on a custom dataset containing `HeteroData` data. I'll add a smaller example to reproduce the issue once I can isolate it.

**Comment by rusty1s** [comment]
That would be great. Thank you!

**Comment by rusty1s** [comment]
Thank you for all the information. I was able to fix it in #4293.
